### PR TITLE
feat: add redis as message broker

### PR DESCRIPTION
**Pull request by 0xthierry**

### CODE REVIEW

**Comment by 0xthierry** [comment]
I'm going to fix the lint issues

**Comment by nerdai** [comment]
> I'm going to fix the lint issues

Thanks! This was super quick given when you first reach out to me to indicate your desire to make this contribute. Will give it a closer look in a few hours time, but initial glance of it looks quite good!

File: pyproject.toml [hunk 1]
```diff
@@ -21,9 +21,11 @@ llama-index-core = "^0.10.50"
 pytest-asyncio = "^0.23.7"
 textual = "^0.70.0"
 aio-pika = {version = "^9.4.2", optional = true}
+redis = {version = "^5.0.7", optional = true}
 
 [tool.poetry.extras]
 rabbitmq = ["aio-pika"]
+redis = ["redis"]
```
Comment (nerdai) [inline pyproject.toml:28]: this is great, can we also add this as a dev dependency as well?

`poetry add redis --group dev`

Comment (nerdai) [inline pyproject.toml:28]: this is great, can we also add this as a dev dependency as well?

poetry add redis --group dev

File: llama_agents/message_queues/redis.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        cls,
+        host: str,
+        port: int = 6379,
+        db: int = 0,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        ssl: bool = False,
+    ) -> "RedisMessageQueue":
+        """Convenience constructor from url params."""
+        scheme = "rediss" if ssl else "redis"
```
Comment (nerdai) [inline llama_agents/message_queues/redis.py:81]: this is great, we should do this for RabbitMQ as well 😊

File: llama_agents/message_queues/redis.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        )
+    return redis.from_url(
+        url,
+    )
+
+
+class Consumer(BaseModel):
+    message_type: str
+    start_consuming_callable: StartConsumingCallable
+    pubsub: Any = None
```
Comment (nerdai) [inline llama_agents/message_queues/redis.py]: I think we have `BaseMessageQueueConsumer` that you can use?

Comment (nerdai) [reply llama_agents/message_queues/redis.py]: I guess we have `pubsub` here. I think we can subclass the base class then?

File: examples/redis/simple-redis-app/docker-compose.yml
Comment (nerdai) [inline examples/redis/simple-redis-app/docker-compose.yml:1]: Not a comment for this docker-compose file specifically, but for the folder. Can we add a small README.md that includes details of how to start the redis server?

**Review by nerdai — COMMENTED** [review]
This is great @0xthierry! Thanks so much. I've left a few comments in my review, but mostly looks good to go.

**Comment by nerdai** [comment]
also, yay unit tests!

File: llama_agents/message_queues/redis.py (hunk 3 above)
Comment (0xthierry) [reply llama_agents/message_queues/redis.py]: I'm not sure about it, I added this consumer to track the consumers and avoid subscribing the same consumer twice / tracking the PubSub instance, it's more to have typing for `self._consumers: Dict[str, Consumer] = {}`.

If I extend `BaseMessageQueueConsumer` I also need to implement the process message and so on.

WDYT?

File: pyproject.toml (hunk 1 above)
Comment (0xthierry) [reply pyproject.toml:28]: [done](https://github.com/run-llama/llama-agents/pull/113/commits/8713ceab562c03b71de6b81accebc50c1584c3f9)!

File: llama_agents/message_queues/redis.py (hunk 3 above)
Comment (nerdai) [reply llama_agents/message_queues/redis.py]: Ah okay, maybe it's just a naming thing then. Perhaps we can call it `ReddisConsumerMetadata`? Leaving it as consumer sort of clashes with name for `BaseMessageQueueConsumer` and could lead to confusion, I feel.

Comment (0xthierry) [reply llama_agents/message_queues/redis.py]: Looks good for me, [done](https://github.com/run-llama/llama-agents/pull/113/commits/c6380ceed67563d7932476203c1cb7d0fda488f0)!

File: examples/redis/simple-redis-app/docker-compose.yml
Comment (0xthierry) [reply examples/redis/simple-redis-app/docker-compose.yml:1]: good point, [done](https://github.com/run-llama/llama-agents/pull/113/commits/d1f8922369f3bd3c5ce3b181fe45bd77d479f09b)!

File: llama_agents/message_queues/redis.py (hunk 2 above)
Comment (0xthierry) [reply llama_agents/message_queues/redis.py:81]: I could add it after the merge of this PR!

**Review by nerdai — APPROVED** [review]
LGTM, @0xthierry! Thanks for working on those changes :)

File: llama_agents/message_queues/redis.py (hunk 2 above)
Comment (nerdai) [reply llama_agents/message_queues/redis.py:81]: that would be wonderful :)

**Pull request merged**